File: stdio-example.js
const setupWasi = require('./setup-wasi')

async function main() {
  // Setup the WASI instance.
  const wasi = await setupWasi('./stdio-example.wasm')

  // Get the functions exported from the WebAssembly
  const {
    writeToStdout,
    writeToStderr,
    callPerror
  } = wasi.instance.exports

  let ptr1 = null
  try {
    ptr1 = wasi.wasiMemoryManager.convertFromString('To stdout\n')
    writeToStdout(ptr1.byteOffset)
  } finally {
    wasi.wasiMemoryManager.free(ptr1)
  }

  try {
    ptr1 = wasi.wasiMemoryManager.convertFromString('To stderr\n')
    writeToStderr(ptr1.byteOffset)
  } finally {
    wasi.wasiMemoryManager.free(ptr1)
  }

  callPerror()
}

main().then(() => console.log('Done'))

File: stdio-example.c
#define __wasi__

#include <stdlib.h>
#include <stdio.h>
#include <locale.h>

int is_locale_initialised = 0;

static void initLocale()
{
    // The locale must be initialised before using
    // multi byte characters.
    is_locale_initialised = 1;
    setlocale(LC_ALL, "");
}

__attribute__((used)) void callPerror(char* ptr)
{
    if (is_locale_initialised == 0)
        initLocale();

    perror("Help!");
}

__attribute__((used)) void writeToStdout(char* ptr)
{
    // printf(ptr);
    fputs(ptr, stdout);
    fflush(stdout);
}

__attribute__((used)) void writeToStderr(char* ptr)
{
    fputs(ptr, stderr);
    fflush(stderr);
}
